File: examples/javascript/tree_1.js
class TreeNode {
  constructor(val) { this.val = val; this.left = null; this.right = null; }
}

class BST {
  constructor() { this.root = null; }
  insert(val) {
    const node = new TreeNode(val);
    if (!this.root) { this.root = node; return; }
    let curr = this.root;
    while (true) {
      if (val < curr.val) {
        if (!curr.left) { curr.left = node; return; }
        curr = curr.left;
      } else {
        if (!curr.right) { curr.right = node; return; }
        curr = curr.right;
      }
    }
  }
  inorder(node = this.root) {
    if (!node) return [];
    return [...this.inorder(node.left), node.val, ...this.inorder(node.right)];
  }
}

const bst = new BST();
[5, 2, 8, 1, 3, 7, 9].forEach(v => bst.insert(v));
console.log(bst.inorder());

File: examples/go/tree_1.go
package main

import "fmt"

type TreeNode struct {
	val   int
	left  *TreeNode
	right *TreeNode
}

func (t *TreeNode) insert(v int) *TreeNode {
	if t == nil {
		return &TreeNode{val: v}
	}
	if v < t.val {
		t.left = t.left.insert(v)
	} else {
		t.right = t.right.insert(v)
	}
	return t
}

func inorder(t *TreeNode) {
	if t == nil {
		return
	}
	inorder(t.left)
	fmt.Println(t.val)
	inorder(t.right)
}

func main() {
	var root *TreeNode
	for _, v := range []int{5, 3, 7, 1, 9} {
		root = root.insert(v)
	}
	inorder(root)
}
